windows/src/desktop/kmconfig/Keyman.System.KMConfigMain.pas

```diff
@@ -0,0 +1,333 @@
+        Exit(False);
+      end;
+    end;
+    if Setting.IsEmpty then
+    begin
+      LogWarning('Setting '+ID+' is already default; not updating.');
+      Exit(True);
+    end;
+    if (Setting.Base.RootKey = HKEY_LOCAL_MACHINE) and not IsAdmin then
+    begin
+      LogError('Setting '+ID+' requires Administrator permissions to update.');
+      Exit(False);
+    end;
+
+    Setting.Reset;
+    if Setting.Modified then
+    begin
+      TKeymanSettingsManager.Save(Settings, IsAdmin);
+      Log('Keyman setting '+ID+' has been reset to default');
+    end
+    else
+    begin
+      LogWarning('Keyman setting '+ID+' was already default; not updating');
+    end;
+  end;
+
+  Result := True;
+end;
+
+class function TKMConfig.SetValue(const ID, Value: string): Boolean;
+var
+  Setting: TKeymanSetting;
+begin
+  Setting := Settings.Find(ID);
+
+  if not Assigned(Setting) then
+  begin
+    // Custom value, do we support it?
+    if ID.StartsWith(CustomKeymanSetting_TSFApp.ID, True) then
+    begin
+      Setting := TKeymanSetting.CreateCustom_TSFApp(ID);
+      Settings.Add(Setting);
+    end
+    else
+    begin
+      LogError(ID+' is not a valid Keyman setting.');
+      Exit(False);
+    end;
+  end;
+
+  if (Setting.Base.RootKey = HKEY_LOCAL_MACHINE) and not IsAdmin then
+  begin
+    LogError('Setting '+ID+' requires Administrator permissions to update.');
+    Exit(False);
+  end;
+
+  case Setting.Base.ValueType of
+    kstString: Setting.ValueStr := Value;
+    kstInteger: Setting.ValueInt := StrToInt(Value);
+  end;
+
+  TKeymanSettingsManager.Save(Settings, IsAdmin);
+  Log('Keyman setting '+ID+' has been updated to '+Value);
+  Result := True;
+end;
+
+class function TKMConfig.Show(const P2: string): Boolean;
+
+  function ShowOneSetting(const ID: string): Boolean;
+  var
+    Setting: TKeymanSetting;
+    v: string;
+  begin
+    Setting := Settings.Find(ID);
+    if not Assigned(Setting) then
+    begin
+      LogError('Setting '+ID+' is not a valid Keyman setting.');
+      Exit(False);
+    end;
+    case Setting.Base.ValueType of
+      kstString: v := Setting.ValueStr;
+      kstInteger: v := Setting.ValueInt.ToString;
+    end;
+    if Setting.IsEmpty
+      then Log(AnsiColor.Green+Setting.Base.ID+AnsiColor.Grey+'='+AnsiColor.Grey+v+AnsiColor.Default)
+      else Log(AnsiColor.Green+Setting.Base.ID+AnsiColor.Grey+'='+AnsiColor.White+v+AnsiColor.Default);
+    Result := True;
+  end;
+
+  procedure ShowAllSettings(ShowAll: Boolean);
+  var
+    Setting: TKeymanSetting;
+  begin
+    for Setting in Settings do
+    begin
+      if ShowAll or not Setting.IsEmpty then
+        ShowOneSetting(Setting.Base.ID);
+    end;
+  end;
+begin
+  Result := True;
+  TKeymanSettingsManager.Load(Settings);
+  if SameText(P2, '-a') then
+    ShowAllSettings(True)
+  else if P2 = '' then
+    ShowAllSettings(False)
+  else
+    Result := ShowOneSetting(P2);
+end;
+
+function DetectColorMode: Boolean;
```

Comment:
This uses the same function as in kmcomp. I haven't attempted to refactor to share code at this point, but we could do so.